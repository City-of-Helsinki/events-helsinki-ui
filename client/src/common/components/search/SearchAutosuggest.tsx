import React, { ChangeEvent, FunctionComponent } from "react";

import { ReactComponent as SearchIcon } from "../../../assets/icons/svg/search.svg";
import { Category as CategoryType } from "../../types";
import Category from "./Category";
import styles from "./searchAutosuggest.module.scss";

interface Props {
  categories: CategoryType[];
  onRemoveCategory: (category: CategoryType) => void;
  placeholder: string;
}

const SearchAutosuggest: FunctionComponent<Props> = ({
  categories,
  onRemoveCategory,
  placeholder
}) => {
  const container = React.useRef<HTMLDivElement | null>(null);
  const input = React.useRef<HTMLInputElement | null>(null);
  const [searchValue, setSearchValue] = React.useState("");

  const handleComponentClick = () => {
    if (input && input.current) {
      // Set focus on input field when clicking container
      input.current.focus();
    }
  };

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(event.target.value);
  };

  React.useEffect(() => {
    const current = container && container.current;

    // Add event listener to component to set focus on input
    if (current) {
      current.addEventListener("click", handleComponentClick);
    }

    // Clean up event listener to prevent memory leaks
    return () => {
      if (current) {
        current.removeEventListener("click", handleComponentClick);
      }
    };
  }, []);

  return (
    <div className={styles.searchAutosuggest} ref={container}>
      <div className={styles.iconWrapper}>
        <SearchIcon />
      </div>
      <div>
        {categories.map(category => {
          return (
            <Category
              key={category.value}
              category={category}
              onRemove={onRemoveCategory}
            />
          );
        })}
      </div>
      <div className={styles.inputWrapper}>
        <input
          ref={input}
          onChange={handleInputChange}
          placeholder={placeholder}
          type="text"
          value={searchValue}
        />
      </div>
    </div>
  );
};

export default SearchAutosuggest;
