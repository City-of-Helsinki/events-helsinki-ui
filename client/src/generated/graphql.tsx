import gql from 'graphql-tag';
import * as ApolloReactCommon from '@apollo/react-common';
import * as ApolloReactHoc from '@apollo/react-hoc';
import * as ApolloReactHooks from '@apollo/react-hooks';
export type Maybe<T> = T | null;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string,
  String: string,
  Boolean: boolean,
  Int: number,
  Float: number,
};

export type EventDetails = {
   __typename?: 'EventDetails',
  id: Scalars['ID'],
  location?: Maybe<Location>,
  keywords: Array<Keyword>,
  superEvent?: Maybe<InternalIdObject>,
  eventStatus?: Maybe<Scalars['String']>,
  externalLinks: Array<ExternalLink>,
  offers: Array<Offer>,
  dataSource?: Maybe<Scalars['String']>,
  publisher?: Maybe<Scalars['ID']>,
  subEvents: Array<InternalIdObject>,
  images: Array<Image>,
  inLanguage: Array<InLanguage>,
  audience: Array<InternalIdObject>,
  createdTime?: Maybe<Scalars['String']>,
  lastModifiedTime?: Maybe<Scalars['String']>,
  datePublished?: Maybe<Scalars['String']>,
  startTime?: Maybe<Scalars['String']>,
  endTime?: Maybe<Scalars['String']>,
  customData?: Maybe<Scalars['String']>,
  audienceMinAge?: Maybe<Scalars['String']>,
  audienceMaxAge?: Maybe<Scalars['String']>,
  superEventType?: Maybe<Scalars['String']>,
  extensionCourse?: Maybe<ExtensionCourse>,
  name: LocalizedObject,
  locationExtraInfo?: Maybe<LocalizedObject>,
  shortDescription?: Maybe<LocalizedObject>,
  provider?: Maybe<LocalizedObject>,
  infoUrl?: Maybe<LocalizedObject>,
  providerContactInfo?: Maybe<Scalars['String']>,
  description?: Maybe<LocalizedObject>,
  internalId?: Maybe<Scalars['String']>,
  internalContext?: Maybe<Scalars['String']>,
  internalType?: Maybe<Scalars['String']>,
};

export type EventListResponse = {
   __typename?: 'EventListResponse',
  meta: Meta,
  data: Array<EventDetails>,
};

export type ExtensionCourse = {
   __typename?: 'ExtensionCourse',
  enrolmentStartTime?: Maybe<Scalars['String']>,
  enrolmentEndTime?: Maybe<Scalars['String']>,
  maximumAttendeeCapacity?: Maybe<Scalars['Int']>,
  minimumAttendeeCapacity?: Maybe<Scalars['Int']>,
  remainingAttendeeCapacity?: Maybe<Scalars['Int']>,
};

export type ExternalLink = {
   __typename?: 'ExternalLink',
  name?: Maybe<Scalars['String']>,
  link?: Maybe<Scalars['String']>,
  language?: Maybe<Scalars['String']>,
};

export type Image = {
   __typename?: 'Image',
  id: Scalars['ID'],
  license?: Maybe<Scalars['String']>,
  createdTime?: Maybe<Scalars['String']>,
  lastModifiedTime?: Maybe<Scalars['String']>,
  name: Scalars['String'],
  url: Scalars['String'],
  cropping?: Maybe<Scalars['String']>,
  photographerName?: Maybe<Scalars['String']>,
  dataSource?: Maybe<Scalars['String']>,
  publisher?: Maybe<Scalars['String']>,
  internalId?: Maybe<Scalars['String']>,
  internalContext?: Maybe<Scalars['String']>,
  internalType?: Maybe<Scalars['String']>,
};

export type InLanguage = {
   __typename?: 'InLanguage',
  id: Scalars['ID'],
  translationAvailable?: Maybe<Scalars['Boolean']>,
  name?: Maybe<LocalizedObject>,
  internalId?: Maybe<Scalars['String']>,
  internalContext?: Maybe<Scalars['String']>,
  internalType?: Maybe<Scalars['String']>,
};

export type InternalIdObject = {
   __typename?: 'InternalIdObject',
  internalId?: Maybe<Scalars['String']>,
};

export type Keyword = {
   __typename?: 'Keyword',
  id: Scalars['String'],
  altLabels: Array<Scalars['String']>,
  createdTime?: Maybe<Scalars['String']>,
  lastModifiedTime: Scalars['String'],
  aggregate?: Maybe<Scalars['Boolean']>,
  deprecated?: Maybe<Scalars['Boolean']>,
  nEvents: Scalars['Int'],
  image?: Maybe<Image>,
  dataSource: Scalars['String'],
  publisher?: Maybe<Scalars['ID']>,
  name: LocalizedObject,
  internalId?: Maybe<Scalars['String']>,
  internalContext?: Maybe<Scalars['String']>,
  internalType?: Maybe<Scalars['String']>,
};

export type LocalizedObject = {
   __typename?: 'LocalizedObject',
  fi?: Maybe<Scalars['String']>,
  sv?: Maybe<Scalars['String']>,
  en?: Maybe<Scalars['String']>,
};

export type Location = {
   __typename?: 'Location',
  id: Scalars['ID'],
  divisions: Array<LocationDivision>,
  createdTime?: Maybe<Scalars['String']>,
  lastModifiedTime?: Maybe<Scalars['String']>,
  customData?: Maybe<Scalars['String']>,
  email?: Maybe<Scalars['String']>,
  contactType?: Maybe<Scalars['String']>,
  addressRegion?: Maybe<Scalars['String']>,
  postalCode?: Maybe<Scalars['String']>,
  postOfficeBoxNum?: Maybe<Scalars['String']>,
  addressCountry?: Maybe<Scalars['String']>,
  deleted?: Maybe<Scalars['Boolean']>,
  nEvents?: Maybe<Scalars['Int']>,
  image?: Maybe<Image>,
  dataSource?: Maybe<Scalars['String']>,
  publisher?: Maybe<Scalars['ID']>,
  parent?: Maybe<Scalars['ID']>,
  replacedBy?: Maybe<Scalars['String']>,
  position?: Maybe<LocationPosition>,
  name?: Maybe<LocalizedObject>,
  description?: Maybe<Scalars['String']>,
  telephone?: Maybe<LocalizedObject>,
  addressLocality?: Maybe<LocalizedObject>,
  streetAddress?: Maybe<LocalizedObject>,
  infoUrl?: Maybe<LocalizedObject>,
  internalId?: Maybe<Scalars['String']>,
  internalContext?: Maybe<Scalars['String']>,
  internalType?: Maybe<Scalars['String']>,
};

export type LocationDivision = {
   __typename?: 'LocationDivision',
  type: Scalars['String'],
  ocdId?: Maybe<Scalars['String']>,
  municipality?: Maybe<Scalars['String']>,
  name?: Maybe<LocalizedObject>,
};

export type LocationPosition = {
   __typename?: 'LocationPosition',
  type: Scalars['String'],
  coordinates: Array<Scalars['Float']>,
};

export type Meta = {
   __typename?: 'Meta',
  count: Scalars['Int'],
  next?: Maybe<Scalars['String']>,
  previous?: Maybe<Scalars['String']>,
};

export type Mutation = {
   __typename?: 'Mutation',
  _empty?: Maybe<Scalars['String']>,
};

export type Offer = {
   __typename?: 'Offer',
  isFree?: Maybe<Scalars['Boolean']>,
  description?: Maybe<LocalizedObject>,
  price?: Maybe<LocalizedObject>,
  infoUrl?: Maybe<LocalizedObject>,
};

export type Query = {
   __typename?: 'Query',
  _empty?: Maybe<Scalars['String']>,
  courseDetails: EventDetails,
  eventDetails: EventDetails,
  eventList: EventListResponse,
};


export type QueryCourseDetailsArgs = {
  id?: Maybe<Scalars['ID']>
};


export type QueryEventDetailsArgs = {
  id?: Maybe<Scalars['ID']>
};

export type Subscription = {
   __typename?: 'Subscription',
  _empty?: Maybe<Scalars['String']>,
};

export type EventDetailsQueryVariables = {
  id: Scalars['ID']
};


export type EventDetailsQuery = (
  { __typename?: 'Query' }
  & { eventDetails: (
    { __typename?: 'EventDetails' }
    & Pick<EventDetails, 'id' | 'endTime' | 'startTime'>
    & { externalLinks: Array<(
      { __typename?: 'ExternalLink' }
      & Pick<ExternalLink, 'name' | 'link'>
    )>, images: Array<(
      { __typename?: 'Image' }
      & Pick<Image, 'id' | 'name' | 'url'>
    )>, inLanguage: Array<(
      { __typename?: 'InLanguage' }
      & { name: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )> }
    )>, keywords: Array<(
      { __typename?: 'Keyword' }
      & Pick<Keyword, 'id' | 'altLabels' | 'createdTime' | 'lastModifiedTime' | 'nEvents' | 'dataSource'>
      & { name: (
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      ) }
    )>, location: Maybe<(
      { __typename?: 'Location' }
      & Pick<Location, 'email' | 'postalCode'>
      & { divisions: Array<(
        { __typename?: 'LocationDivision' }
        & Pick<LocationDivision, 'type'>
        & { name: Maybe<(
          { __typename?: 'LocalizedObject' }
          & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
        )> }
      )>, infoUrl: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )>, name: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'en' | 'sv'>
      )>, addressLocality: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )>, streetAddress: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )>, position: Maybe<(
        { __typename?: 'LocationPosition' }
        & Pick<LocationPosition, 'coordinates'>
      )>, telephone: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )> }
    )>, offers: Array<(
      { __typename?: 'Offer' }
      & Pick<Offer, 'isFree'>
      & { price: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )>, description: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )>, infoUrl: Maybe<(
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
      )> }
    )>, name: (
      { __typename?: 'LocalizedObject' }
      & Pick<LocalizedObject, 'fi' | 'en' | 'sv'>
    ), description: Maybe<(
      { __typename?: 'LocalizedObject' }
      & Pick<LocalizedObject, 'fi' | 'en' | 'sv'>
    )>, shortDescription: Maybe<(
      { __typename?: 'LocalizedObject' }
      & Pick<LocalizedObject, 'fi' | 'en' | 'sv'>
    )>, infoUrl: Maybe<(
      { __typename?: 'LocalizedObject' }
      & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
    )> }
  ) }
);

export type EventListQueryVariables = {};


export type EventListQuery = (
  { __typename?: 'Query' }
  & { eventList: (
    { __typename?: 'EventListResponse' }
    & { meta: (
      { __typename?: 'Meta' }
      & Pick<Meta, 'count' | 'next' | 'previous'>
    ), data: Array<(
      { __typename?: 'EventDetails' }
      & Pick<EventDetails, 'id' | 'startTime' | 'endTime'>
      & { images: Array<(
        { __typename?: 'Image' }
        & Pick<Image, 'id' | 'name' | 'url'>
      )>, location: Maybe<(
        { __typename?: 'Location' }
        & { divisions: Array<(
          { __typename?: 'LocationDivision' }
          & Pick<LocationDivision, 'type'>
          & { name: Maybe<(
            { __typename?: 'LocalizedObject' }
            & Pick<LocalizedObject, 'fi' | 'en' | 'sv'>
          )> }
        )>, name: Maybe<(
          { __typename?: 'LocalizedObject' }
          & Pick<LocalizedObject, 'fi' | 'en' | 'sv'>
        )>, addressLocality: Maybe<(
          { __typename?: 'LocalizedObject' }
          & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
        )>, streetAddress: Maybe<(
          { __typename?: 'LocalizedObject' }
          & Pick<LocalizedObject, 'fi' | 'sv' | 'en'>
        )> }
      )>, name: (
        { __typename?: 'LocalizedObject' }
        & Pick<LocalizedObject, 'fi' | 'en' | 'sv'>
      ) }
    )> }
  ) }
);


export const EventDetailsDocument = gql`
    query EventDetails($id: ID!) {
  eventDetails(id: $id) {
    id
    externalLinks {
      name
      link
    }
    images {
      id
      name
      url
    }
    inLanguage {
      name {
        fi
        sv
        en
      }
    }
    keywords {
      id
      altLabels
      createdTime
      lastModifiedTime
      nEvents
      dataSource
      name {
        fi
        sv
        en
      }
    }
    location {
      divisions {
        type
        name {
          fi
          sv
          en
        }
      }
      email
      infoUrl {
        fi
        sv
        en
      }
      name {
        fi
        en
        sv
      }
      addressLocality {
        fi
        sv
        en
      }
      streetAddress {
        fi
        sv
        en
      }
      postalCode
      position {
        coordinates
      }
      telephone {
        fi
        sv
        en
      }
    }
    offers {
      isFree
      price {
        fi
        sv
        en
      }
      description {
        fi
        sv
        en
      }
      infoUrl {
        fi
        sv
        en
      }
    }
    name {
      fi
      en
      sv
    }
    description {
      fi
      en
      sv
    }
    shortDescription {
      fi
      en
      sv
    }
    endTime
    startTime
    infoUrl {
      fi
      sv
      en
    }
  }
}
    `;
export type EventDetailsProps<TChildProps = {}> = ApolloReactHoc.DataProps<EventDetailsQuery, EventDetailsQueryVariables> | TChildProps;
export function withEventDetails<TProps, TChildProps = {}>(operationOptions?: ApolloReactHoc.OperationOption<
  TProps,
  EventDetailsQuery,
  EventDetailsQueryVariables,
  EventDetailsProps<TChildProps>>) {
    return ApolloReactHoc.withQuery<TProps, EventDetailsQuery, EventDetailsQueryVariables, EventDetailsProps<TChildProps>>(EventDetailsDocument, {
      alias: 'eventDetails',
      ...operationOptions
    });
};

/**
 * __useEventDetailsQuery__
 *
 * To run a query within a React component, call `useEventDetailsQuery` and pass it any options that fit your needs.
 * When your component renders, `useEventDetailsQuery` returns an object from Apollo Client that contains loading, error, and data properties 
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useEventDetailsQuery({
 *   variables: {
 *      id: // value for 'id'
 *   },
 * });
 */
export function useEventDetailsQuery(baseOptions?: ApolloReactHooks.QueryHookOptions<EventDetailsQuery, EventDetailsQueryVariables>) {
        return ApolloReactHooks.useQuery<EventDetailsQuery, EventDetailsQueryVariables>(EventDetailsDocument, baseOptions);
      }
export function useEventDetailsLazyQuery(baseOptions?: ApolloReactHooks.LazyQueryHookOptions<EventDetailsQuery, EventDetailsQueryVariables>) {
          return ApolloReactHooks.useLazyQuery<EventDetailsQuery, EventDetailsQueryVariables>(EventDetailsDocument, baseOptions);
        }
export type EventDetailsQueryHookResult = ReturnType<typeof useEventDetailsQuery>;
export type EventDetailsLazyQueryHookResult = ReturnType<typeof useEventDetailsLazyQuery>;
export type EventDetailsQueryResult = ApolloReactCommon.QueryResult<EventDetailsQuery, EventDetailsQueryVariables>;
export const EventListDocument = gql`
    query EventList {
  eventList {
    meta {
      count
      next
      previous
    }
    data {
      id
      images {
        id
        name
        url
      }
      location {
        divisions {
          type
          name {
            fi
            en
            sv
          }
        }
        name {
          fi
          en
          sv
        }
        addressLocality {
          fi
          sv
          en
        }
        streetAddress {
          fi
          sv
          en
        }
      }
      name {
        fi
        en
        sv
      }
      startTime
      endTime
    }
  }
}
    `;
export type EventListProps<TChildProps = {}> = ApolloReactHoc.DataProps<EventListQuery, EventListQueryVariables> | TChildProps;
export function withEventList<TProps, TChildProps = {}>(operationOptions?: ApolloReactHoc.OperationOption<
  TProps,
  EventListQuery,
  EventListQueryVariables,
  EventListProps<TChildProps>>) {
    return ApolloReactHoc.withQuery<TProps, EventListQuery, EventListQueryVariables, EventListProps<TChildProps>>(EventListDocument, {
      alias: 'eventList',
      ...operationOptions
    });
};

/**
 * __useEventListQuery__
 *
 * To run a query within a React component, call `useEventListQuery` and pass it any options that fit your needs.
 * When your component renders, `useEventListQuery` returns an object from Apollo Client that contains loading, error, and data properties 
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useEventListQuery({
 *   variables: {
 *   },
 * });
 */
export function useEventListQuery(baseOptions?: ApolloReactHooks.QueryHookOptions<EventListQuery, EventListQueryVariables>) {
        return ApolloReactHooks.useQuery<EventListQuery, EventListQueryVariables>(EventListDocument, baseOptions);
      }
export function useEventListLazyQuery(baseOptions?: ApolloReactHooks.LazyQueryHookOptions<EventListQuery, EventListQueryVariables>) {
          return ApolloReactHooks.useLazyQuery<EventListQuery, EventListQueryVariables>(EventListDocument, baseOptions);
        }
export type EventListQueryHookResult = ReturnType<typeof useEventListQuery>;
export type EventListLazyQueryHookResult = ReturnType<typeof useEventListLazyQuery>;
export type EventListQueryResult = ApolloReactCommon.QueryResult<EventListQuery, EventListQueryVariables>;