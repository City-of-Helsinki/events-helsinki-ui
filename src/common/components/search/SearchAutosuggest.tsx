import { IconSearch } from 'hds-react';
import React, { ChangeEvent } from 'react';

import {
  AUTOSUGGEST_KEYWORD_BLACK_LIST,
  AUTOSUGGEST_TYPES,
} from '../../../constants';
import {
  LinkedEventsSource,
  useKeywordListQuery,
} from '../../../generated/graphql';
import useDebounce from '../../../hooks/useDebounce';
import useKeyboardNavigation from '../../../hooks/useDropdownKeyboardNavigation';
import useLocale from '../../../hooks/useLocale';
import getLocalisedString from '../../../util/getLocalisedString';
import { AutosuggestMenuOption } from '../../types';
import AutosuggestMenu from './AutosuggestMenu';
import styles from './searchAutosuggest.module.scss';

export interface SearchAutosuggestProps {
  name: string;
  onChangeSearchValue: (value: string) => void;
  onOptionClick: (item: AutosuggestMenuOption) => void;
  placeholder: string;
  searchValue: string;
  source?: LinkedEventsSource;
}

const SearchAutosuggest: React.FC<SearchAutosuggestProps> = ({
  name,
  onOptionClick,
  onChangeSearchValue,
  placeholder,
  searchValue,
  source,
}) => {
  const locale = useLocale();
  const container = React.useRef<HTMLDivElement | null>(null);
  const input = React.useRef<HTMLInputElement | null>(null);
  const [isMenuOpen, setIsMenuOpen] = React.useState(false);
  const internalInputValue = useDebounce(searchValue, 300);

  const { data: keywordsData, loading: loadingKeywords } = useKeywordListQuery({
    skip: !internalInputValue,
    variables: {
      hasUpcomingEvents: true,
      pageSize: 5,
      text: internalInputValue,
      source,
    },
  });

  const [autoSuggestItems, setAutoSuggestItems] = React.useState<
    AutosuggestMenuOption[]
  >([]);

  const ensureMenuIsClosed = React.useCallback(() => {
    if (isMenuOpen) {
      setIsMenuOpen(false);
    }
  }, [isMenuOpen]);

  const ensureMenuIsOpen = React.useCallback(() => {
    if (!isMenuOpen) {
      setIsMenuOpen(true);
    }
  }, [isMenuOpen]);

  const {
    focusedIndex,
    setup: setupKeyboardNav,
    teardown: teardownKeyboardNav,
  } = useKeyboardNavigation({
    container: container,
    listLength: autoSuggestItems.length,
    onKeyDown: (event: KeyboardEvent) => {
      switch (event.key) {
        case 'ArrowUp':
        case 'ArrowDown':
          ensureMenuIsOpen();
          break;
        case 'Escape':
          handleCloseMenu();
          break;
        case 'Enter':
          const selectedItem = autoSuggestItems[focusedIndex];

          if (selectedItem) {
            handleMenuOptionClick(selectedItem);
          } else {
            // Search by text if no option is selected
            handleMenuOptionClick({
              text: searchValue,
              type: AUTOSUGGEST_TYPES.TEXT,
              value: searchValue,
            });
          }
          break;
        case 'Tab':
          ensureMenuIsClosed();
      }
    },
  });

  React.useEffect(() => {
    if (loadingKeywords) return;

    const items: AutosuggestMenuOption[] = [];
    const textItem = {
      text: internalInputValue,
      type: AUTOSUGGEST_TYPES.TEXT,
      value: internalInputValue,
    };

    items.push(textItem);

    items.push(
      ...(keywordsData?.keywordList.data
        .filter((keyword) => {
          const name = getLocalisedString(keyword.name, locale).toLowerCase();
          return (
            !AUTOSUGGEST_KEYWORD_BLACK_LIST.includes(keyword.id || '') &&
            name &&
            name !== textItem.text.toLowerCase()
          );
        })
        .map((keyword) => ({
          text: getLocalisedString(keyword.name, locale),
          type: AUTOSUGGEST_TYPES.KEYWORD,
          value: keyword.id || '',
        })) || [])
    );

    setAutoSuggestItems(items.filter((item) => item.text));
  }, [internalInputValue, keywordsData, loadingKeywords, locale]);

  const handleCloseMenu = React.useCallback(() => {
    setFocusToInput();

    ensureMenuIsClosed();
  }, [ensureMenuIsClosed]);

  const handleMenuOptionClick = React.useCallback(
    (option: AutosuggestMenuOption) => {
      onOptionClick(option);
      handleCloseMenu();
    },
    [handleCloseMenu, onOptionClick]
  );

  const setFocusToInput = () => {
    input.current?.focus();
  };

  const handleComponentClick = () => {
    setFocusToInput();
  };

  const onDocumentClick = (event: MouseEvent) => {
    const target = event.target;
    const current = container.current;

    // Close menu when clicking outside of the component
    if (!(target instanceof Node && current?.contains(target))) {
      setIsMenuOpen(false);
    }
  };

  const onDocumentFocusin = React.useCallback(
    (event: FocusEvent) => {
      const target = event.target;
      const current = container.current;

      if (!(target instanceof Node && current?.contains(target))) {
        ensureMenuIsClosed();
      }
    },
    [ensureMenuIsClosed]
  );

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;

    onChangeSearchValue(newValue);

    ensureMenuIsOpen();
  };

  const handleInputFocus = () => {
    ensureMenuIsOpen();
  };

  React.useEffect(() => {
    setupKeyboardNav();
    document.addEventListener('click', onDocumentClick);
    document.addEventListener('focusin', onDocumentFocusin);
    // Clean up event listener to prevent memory leaks
    return () => {
      teardownKeyboardNav();
      document.removeEventListener('click', onDocumentClick);
      document.removeEventListener('focusin', onDocumentFocusin);
    };
  }, [onDocumentFocusin, setupKeyboardNav, teardownKeyboardNav]);

  return (
    <div
      onClick={handleComponentClick}
      className={styles.searchAutosuggest}
      ref={container}
    >
      <div className={styles.iconWrapper}>
        <IconSearch size="s" aria-hidden />
      </div>
      <div className={styles.inputWrapper}>
        <input
          ref={input}
          id={name}
          name={name}
          onChange={handleInputChange}
          onFocus={handleInputFocus}
          placeholder={placeholder}
          type="text"
          value={searchValue}
        />
      </div>
      <AutosuggestMenu
        focusedOption={focusedIndex}
        isOpen={!!searchValue && isMenuOpen}
        onClose={handleCloseMenu}
        onOptionClick={handleMenuOptionClick}
        options={autoSuggestItems}
      />
    </div>
  );
};

export default SearchAutosuggest;
