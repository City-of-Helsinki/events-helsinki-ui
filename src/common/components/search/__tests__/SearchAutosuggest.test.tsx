import { MockedProvider } from '@apollo/react-testing';
import * as React from 'react';
import renderer from 'react-test-renderer';

import mockKeyword from '../../../../domain/keyword/__mocks__/keyword';
import { KeywordListDocument } from '../../../../generated/graphql';
import SearchAutosuggest from '../SearchAutosuggest';

const mocks = [
  {
    request: {
      query: KeywordListDocument,
      variables: {
        hasUpcomingEvents: true,
        pageSize: 5,
        text: 'search value',
      },
    },
    result: {
      data: {
        keywordList: {
          __typename: 'KeywordListResponse',
          data: [mockKeyword],
        },
      },
    },
  },
];

test('SearchAutosuggest matches snapshot', () => {
  const component = renderer.create(
    <MockedProvider mocks={mocks} addTypename={true}>
      <SearchAutosuggest
        categories={[{ text: 'bar', value: 'foo' }]}
        name="search"
        onRemoveCategory={category => {}}
        onChangeSearchValue={value => {}}
        onOptionClick={() => {}}
        placeholder="Lorem ipsum"
        searchValue={'search value'}
      />
    </MockedProvider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

export {};
