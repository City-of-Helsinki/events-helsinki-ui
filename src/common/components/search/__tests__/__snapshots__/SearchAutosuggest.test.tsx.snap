// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="searchAutosuggest"
>
  <div
    class="iconWrapper"
  >
    <svg
      class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <path
          d="M0 0h24v24H0z"
        />
        <path
          d="M15 1a8 8 0 11-4.798 14.402l-6.401 6.4-1.591-1.59 6.398-6.4A8 8 0 0115 1zm0 2a6 6 0 100 12 6 6 0 000-12z"
          fill="currentColor"
        />
      </g>
    </svg>
  </div>
  <div
    class="inputWrapper"
  >
    <input
      id="search"
      name="search"
      placeholder="Placeholder text"
      type="text"
      value="jaz"
    />
  </div>
</div>
`;
